lib/matplotlib/collections.py

```diff
@@ -1989,23 +2019,75 @@ class QuadMesh(Collection):
     vertex at mesh coordinates (0, 0), then the one at (0, 1), then at (0, 2)
     .. (0, meshWidth), (1, 0), (1, 1), and so on.
 
-    *shading* may be 'flat', or 'gouraud'
     """
-    def __init__(self, meshWidth, meshHeight, coordinates,
-                 antialiased=True, shading='flat', **kwargs):
-        super().__init__(**kwargs)
+    def __init__(self, *args, **kwargs):
+        # signature deprecation since="3.5": Change to new_sig after the
+        # deprecation has expired. Also remove setting __init__.__signature__.
+        old_sig = inspect.signature(
+            lambda meshWidth, meshHeight, coordinates,
+                   antialiased=True, shading='flat', **kwargs: None)
+        new_sig = inspect.signature(
+            lambda coordinates, *,
+                   antialiased=True, shading='flat', **kwargs: None)
+        try:
+            bound = new_sig.bind(*args, **kwargs)
+        except TypeError as exc:
+            try:
+                print(args, kwargs)
+                bound = old_sig.bind(*args, **kwargs)
+            except TypeError:
+                raise exc  # the new_sig binding error makes more sense
+                # user passed in args using the old call convention
+            else:
+                bound.apply_defaults()
```

Comment:
Sorry for now realizing this earlier, but a simpler solution would be to do something like
```python
try:
    # The following raises a TypeError iif the args don't match.
    w, h, c, a, s, k = (
        lambda meshWidth, meshHeight, coordinates, antialiased=True, shading=False, **kwargs:
        (meshWidth, meshHeight, coordinates, antialiased, shading, kwargs))(*args, **kwargs)
except TypeError as exc:
    # If the following raises a TypeError (i.e. args don't match), just let it propagate.
    c, a, s = (
        lambda coordinates, antialiased=True, shading=False, **kwargs:
        (coordinates, antialiased, shading, kwargs))(*args, **kwargs)
else:  # The old signature matched.
    warn_deprecated(...)
    c = np.asarray(c, float).reshape((h + 1, w + 1, 2))
```
avoiding the need to `apply_defaults()` and to extract args from `bound`.